Source/WebKit/GPUProcess/graphics/RemoteImageBufferSet.cpp

```diff
@@ -0,0 +1,258 @@
+ */
+
+#include "config.h"
+#include "RemoteImageBufferSet.h"
+
+#include "ImageBufferBackendHandleSharing.h"
+#include "Logging.h"
+#include "RemoteImageBufferSetMessages.h"
+#include "RemoteRenderingBackend.h"
+#include <WebCore/PlatformCALayer.h>
+
+#if ENABLE(GPU_PROCESS)
+
+#define MESSAGE_CHECK(assertion, backend, message) do { \
+    if (UNLIKELY(!(assertion))) { \
+        backend.terminateWebProcess(message); \
+        return; \
+    } \
+} while (0)
+
+namespace WebKit {
+
+Ref<RemoteImageBufferSet> RemoteImageBufferSet::create(RemoteImageBufferSetIdentifier identifier, RemoteRenderingBackend& backend)
+{
+    auto instance = adoptRef(*new RemoteImageBufferSet(identifier, backend));
+    instance->startListeningForIPC();
+    return instance;
+}
+
+RemoteImageBufferSet::RemoteImageBufferSet(RemoteImageBufferSetIdentifier identifier, RemoteRenderingBackend& backend)
+    : m_backend(&backend)
+    , m_identifier(identifier)
+{
+}
+
+RemoteImageBufferSet::~RemoteImageBufferSet() = default;
+
+void RemoteImageBufferSet::startListeningForIPC()
+{
+    m_backend->streamConnection().startReceivingMessages(*this, Messages::RemoteImageBufferSet::messageReceiverName(), m_identifier.toUInt64());
+}
+
+void RemoteImageBufferSet::stopListeningForIPC()
+{
+    if (auto backend = std::exchange(m_backend, { }))
+        backend->streamConnection().stopReceivingMessages(Messages::RemoteImageBufferSet::messageReceiverName(), m_identifier.toUInt64());
+}
+
+IPC::StreamConnectionWorkQueue& RemoteImageBufferSet::workQueue() const
+{
+    return m_backend->workQueue();
+}
+
+void RemoteImageBufferSet::updateConfiguration(const FloatSize& logicalSize, RenderingMode renderingMode, float resolutionScale, const DestinationColorSpace& colorSpace, PixelFormat pixelFormat)
+{
+    m_logicalSize = logicalSize;
+    m_renderingMode = renderingMode;
+    m_resolutionScale = resolutionScale;
+    m_colorSpace = colorSpace;
+    m_pixelFormat = pixelFormat;
+    m_frontBuffer = m_backBuffer = m_secondaryBackBuffer = nullptr;
+}
+
+void RemoteImageBufferSet::setFlushSignal(IPC::Signal&& signal)
+{
+    m_flushSignal = WTFMove(signal);
+}
+
+void RemoteImageBufferSet::flush()
+{
+    if (m_frontBuffer) {
+        ASSERT(m_flushSignal);
+        m_frontBuffer->flushDrawingContext();
+        m_flushSignal->signal();
+    }
+}
+
+// This is the GPU Process version of RemoteLayerBackingStore::prepareBuffers().
+void RemoteImageBufferSet::ensureBufferForDisplay(const ImageBufferSetPrepareBufferForDisplayInputData& inputData, ImageBufferSetPrepareBufferForDisplayOutputData& outputData)
+{
+    assertIsCurrent(workQueue());
+    auto bufferIdentifier = [](RefPtr<WebCore::ImageBuffer> buffer) -> std::optional<ProcessQualified<RenderingResourceIdentifier>> {
+        if (!buffer)
+            return std::nullopt;
+        return ProcessQualified(buffer->renderingResourceIdentifier(), Process::identifier());
```

Comment:
I think this is that you replace the "generation" with process identifier?
I'm not 100% if we already transfer GPUP identifier to WP. However, I don't think it makes sense?
If we transfer 10000 image buffers, we keep on transferring the very same 64-bit id on an on, which could be somehow done differently without transferring this?
BTW This refactoring could have been made separately, so it'd be easier to review and land...